gst/nnstreamer/nnstreamer_subplugin.c

```diff
@@ -0,0 +1,206 @@
+ *
+ */
+
+#include <dlfcn.h>
+#include <glib.h>
+#include <gmodule.h>
+#include <gst/gstinfo.h>
+#include "nnstreamer_subplugin.h"
+#include "nnstreamer_conf.h"
+
+typedef struct
+{
+  char *name; /**< The name of subplugin */
+  const void *data; /**< subplugin specific data forwarded from the subplugin */
+  void *handle; /**< dlopen'ed handle */
+} subpluginData;
+
+static GHashTable *subplugins[NNS_SUBPLUGIN_END] = { 0 };
+
+G_LOCK_DEFINE_STATIC (splock);
+
+/**
+ * @brief Private function (hash func) for GHashTable. Type = GHashFunc
+ */
+static guint
+_sphash (gconstpointer key)
+{
+  const subpluginData *data = key;
+  return g_str_hash (data->name);
+}
+
+/**
+ * @brief Private function (key comp func) for GHashTable. Type = GEqualFunc
+ */
+static gboolean
+_strequal (gconstpointer _a, gconstpointer _b)
+{
+  const subpluginData *a = _a, *b = _b;
+  return (0 == g_strcmp0 (a->name, b->name));
+}
+
+/** @brief Public function defined in the header */
+const void *
+get_subplugin (subpluginType type, const char *name)
+{
+  GHashTable *table;
+  subpluginData *data;
+  subpluginData key;
+
+  if (subplugins[type] == NULL)
+    subplugins[type] = g_hash_table_new (_sphash, _strequal);
+
+  table = subplugins[type];
+
+  key.name = g_strdup (name);
+
+  data = g_hash_table_lookup (table, &key);
+
+
+  if (data == NULL) {
+    /* Search and register if found with the conf */
+    const gchar *fullpath = nnsconf_get_fullpath (name, type);
+    void *handle;
+    char *dlsym_error;
+    nnstreamer_subplugin_data *nsdata;
+
+    if (fullpath == NULL)
+      goto error;               /* No Such Thing !!! */
+
+    handle = dlopen (fullpath, RTLD_NOW);
+    if (NULL == handle) {
+      GST_ERROR ("Cannot dlopen %s (%s).", name, fullpath);
+      goto error;
+    }
+
+    nsdata = (nnstreamer_subplugin_data *)
+        dlsym (handle, "nnstreamer_subplugin");
+    dlsym_error = dlerror ();
+
+    if (NULL == nsdata) {
+      GST_ERROR ("nnstreamer_subplugin does not exists in %s (%s)", name,
+          fullpath);
+      dlclose (handle);
+      goto error;
+    }
+    if (dlsym_error) {
+      GST_ERROR ("Loading nnstreamer_subplugin in %s (%s) incurs: %s", name,
+          fullpath, dlsym_error);
+      dlclose (handle);
+      goto error;
+    }
+
+    if (nsdata->checker != NNS_SUBPLUGIN_CHECKER) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (first bytes)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    if (nsdata->type != type) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (type mismatch)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    if (g_strcmp0 (nsdata->name, name)) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (name mismatch)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    g_assert (TRUE == register_subplugin (type, name, data));
```

Comment:
register_subplugin (type, name, data)
at this point, data is null, is it correct?

Reply:
Fixed. Thanks.